### PR TITLE
Record/retrieve all sample scans

### PR DESCRIPTION
**Pull request by AmandaBirmingham**
Ready for a real code review :) Moves scan of a barcode out into a new barcode_scans table, updates api to store new records to this table when scanning is done and to retrieve all scans when getting diagnostic info for a barcode.

GUI to this new functionality is in PR #39 in biocore/microsetta-admin.

### CODE REVIEW

File: microsetta_private_api/repo/sample_repo.py [hunk 1]
```diff
@@ -76,74 +54,86 @@ def _retrieve_projects(self, sample_barcode):
...
+    def _create_sample_obj(self, sample_row):
+        if sample_row is None:
+            return None
+
+        sample_barcode = sample_row[5]
+        sample_projects = self._retrieve_projects(sample_barcode)
+
+        return Sample.from_db(*sample_row, sample_projects)
+
+    # TODO: I'm still not entirely happy with the linking between samples and
```
Comment (wasade) [inline microsetta_private_api/repo/sample_repo.py:66]: Completely agree, I think we really should figure out how to streamline this

File: microsetta_private_api/admin/admin_impl.py [hunk 2]
```diff
@@ -45,12 +44,13 @@ def search_email(token_info, email):
 def scan_barcode(token_info, sample_barcode, body):
     validate_admin_access(token_info)
 
+    scan_id = None
```
Comment (dhakim87) [inline microsetta_private_api/admin/admin_impl.py:47]: Not a change request.  Just a question I've wondered.  This is the standard style in Java/C#, but do python's variable scopes work the same way?  My Pycharm frequently highlights lines like this that initialize a None that can't actually be used as being unnecessary.

File: microsetta_private_api/db/patches/0067.sql [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+-- account by way of a source, and has collection info.  Note that
+-- collection date is used as a proxy for collection info
+-- because all kinds of sources require collection_date for
+-- a sample (whereas, e.g., environmental sources don't require
+-- site_sampled, etc).
+UPDATE barcodes.barcode
+SET sample_status = 'valid'
+FROM ag.ag_kit_barcodes, ag.source, ag.account
+WHERE barcode.barcode = ag_kit_barcodes.barcode
+AND ag_kit_barcodes.source_id = ag.source.id
```
[the remark this answers is not in the document]
Comment (dhakim87) [reply microsetta_private_api/db/patches/0067.sql:18]: Wow, I've never seen this syntax before.  
StackOverflow says line 16 does an implicit inner join, but I had no idea such syntax existed.  

EXPLAIN says:

Update on barcode  (cost=3107.70..6125.22 rows=992 width=118)
   ->  Nested Loop  (cost=3107.70..6125.22 rows=992 width=118)
         ->  Hash Join  (cost=3107.42..5588.38 rows=992 width=128)
               Hash Cond: (ag_kit_barcodes.source_id = source.id)
               ->  Hash Join  (cost=2313.92..4774.80 rows=2709 width=122)
                     Hash Cond: ((ag_kit_barcodes.barcode)::text = (barcode.barcode)::text)
                     ->  Seq Scan on ag_kit_barcodes  (cost=0.00..2367.61 rows=17648 width=32)
                           Filter: (site_sampled IS NOT NULL)
                     ->  Hash  (cost=2177.73..2177.73 rows=10895 width=100)
                           ->  Seq Scan on barcode  (cost=0.00..2177.73 rows=10895 width=100)
                                 Filter: ((sample_status IS NULL) AND (scan_date IS NOT NULL))
               ->  Hash  (cost=619.89..619.89 rows=13889 width=38)
                     ->  Seq Scan on source  (cost=0.00..619.89 rows=13889 width=38)
         ->  Index Scan using account_pkey on account  (cost=0.29..0.53 rows=1 width=22)
               Index Cond: (id = source.account_id)
(15 rows)

File: microsetta_private_api/admin/admin_impl.py (hunk 2 above)
Comment (wasade) [reply microsetta_private_api/admin/admin_impl.py:47]: In this case I think it's unnecessary as, unless i'm mistaken, the context manager is assured to enter and exit. I don't believe there is a conditional flow here. If the `with` were a try/except, then there is a possibility in which the use of `scan_id` on like 53 would trigger an `UnboundLocalError` (or `NameError`, I don't recall exact semantics)

File: microsetta_private_api/repo/admin_repo.py [hunk 4]
```diff
@@ -46,6 +45,9 @@ def _get_ids_relevant_to_barcode(self, sample_barcode):
             return cur.fetchone()
 
     def retrieve_diagnostics_by_barcode(self, sample_barcode, grab_kit=True):
+        def _rows_to_dicts_list(rows):
+            return [{k: v for k, v in x.items()} for x in rows]
```
[the remark this answers is not in the document]
Comment (dhakim87) [reply microsetta_private_api/repo/admin_repo.py:49]: Haha, you moved my code and got a comment on it :D

[the remark this answers is not in the document]
Comment (AmandaBirmingham) [reply microsetta_private_api/repo/admin_repo.py:49]: Didn't know you could do that :) Done.

[the remark this answers is not in the document]
Comment (wasade) [reply microsetta_private_api/repo/admin_repo.py]: I mean, I'd verify it works of course... :)

File: microsetta_private_api/db/patches/0067.sql (hunk 3 above)
[the remark this answers is not in the document]
Comment (AmandaBirmingham) [reply microsetta_private_api/db/patches/0067.sql:18]: I do not know enough about postgresql to know why this raises concerns :( I just googled "postgresql update join" and followed examples like this one.  Happy to be educated on the preferred approach.
Screen Shot 2020-07-17 at 2 56 47 PM

File: microsetta_private_api/repo/admin_repo.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AmandaBirmingham) [reply microsetta_private_api/repo/admin_repo.py]: Mais bien sûr ! :D

**Pull request merged**